Bootloader/Bootloader/UnitTests/Mcal_ut/Dio_ut/Dio_ut.c

```diff
@@ -1,6 +1,173 @@
+	Dio_DDRC = 0;
+	
+	Dio_PIND = 0;
+	Dio_PORTD	= 0;
+	Dio_DDRD = 0;
+}
+
+static void Test_SetPinDirection_ValidInput()
+{
+	ResetAllRegisters();
+	boolean Loc_Return = Dio_SetPinDirection(0,2,0);
+	UT_ASSERT_EQ("SetPinDirection_ValidInput", "Dio", Loc_Return, 1,"Correct Return status" ,"Wrong Return status");
+	UT_ASSERT_EQ("SetPinDirection_ValidInput", "Dio", Dio_DDRA>>2, 0,"Correct DDRA_Pin2 data" ,"Wrong DDRA_Pin2 data");
+}
+
+static void Test_SetPinDirection_InValidInput()
+{
+	ResetAllRegisters();
+	boolean Loc_Return = Dio_SetPinDirection(3,9,0);
+	UT_ASSERT_EQ("SetPinDirection_InValidInput", "Dio", Loc_Return, 0,"Correct Return status" ,"Wrong Return status");
+	Loc_Return = Dio_SetPinDirection(9,4,0);
+	UT_ASSERT_EQ("SetPinDirection_InValidInput", "Dio", Loc_Return, 0,"Correct Return status" ,"Wrong Return status");
+}
+
+static void Test_SetPortDirection_ValidInput()
+{
+	ResetAllRegisters();
+	boolean Loc_Return = Dio_SetPortDirection(1,1);
+	UT_ASSERT_EQ("SetPortDirection_ValidInput", "Dio", Loc_Return, 1,"Correct Return status" ,"Wrong Return status");
+	UT_ASSERT_EQ("SetPortDirection_ValidInput", "Dio", Dio_DDRB, 1,"Correct DDRB data" ,"Wrong DDRB data");
+}
+
+static void Test_SetPortDirection_InValidInput()
+{
+	ResetAllRegisters();
+	boolean Loc_Return = Dio_SetPortDirection(9,0);
+	UT_ASSERT_EQ("SetPinDirection_InValidInput", "Dio", Loc_Return, 0,"Correct Return status" ,"Wrong Return status");
+}
+
+static void Test_WriteBit_ValidInput()
+{
+	ResetAllRegisters();
+	boolean Loc_Return = Dio_WritePin(3,4,1);
+	UT_ASSERT_EQ("WriteBit_ValidInput", "Dio", Loc_Return, 1,"Correct Return status" ,"Wrong Return status");
+	UT_ASSERT_EQ("WriteBit_ValidInput", "Dio", Dio_PORTD>>4, 1,"Correct Pin data" ,"Wrong pin data");
+}
+
+static void Test_ReadBit_ValidInput()
+{
+	ResetAllRegisters();
+	Dio_PINC = 0b10000000;
+	uint8 Loc_ReturnByte = 0;
+	boolean Loc_Return = Dio_ReadPin(2,7, &Loc_ReturnByte);
+	UT_ASSERT_EQ("ReadBit_ValidInput", "Dio", Loc_Return, 1,"Correct Return status" ,"Wrong Return status");
+	UT_ASSERT_EQ("ReadBit_ValidInput", "Dio", Loc_ReturnByte, 1,"Correct PORTC_Pin7 data" ,"Wrong PORTC_Pin7 data");
+} 
+
+static void Test_WriteBit_InValidInput()
+{
+	ResetAllRegisters();
+	boolean Loc_Return = Dio_WritePin(9,4,1);
+	UT_ASSERT_EQ("WriteBit_InValidInput", "Dio", Loc_Return, 0,"Correct Return status" ,"Wrong Return status");
+	Loc_Return = Dio_WritePin(1,9,1);
+	UT_ASSERT_EQ("WriteBit_InValidInput", "Dio", Loc_Return, 0,"Correct Return status" ,"Wrong Return status");
+	
+}
+
+static void Test_ReadBit_InValidInput()
+{
+	ResetAllRegisters();
+	
+	Dio_PINC = 0b10000000;
+	uint8 Loc_ReturnByte = 0;
+	boolean Loc_Return = Dio_ReadPin(2,8, &Loc_ReturnByte);
+	UT_ASSERT_EQ("ReadBit_InValidInput", "Dio", Loc_Return, 0,"Correct Return status" ,"Wrong Return status");
+	Loc_Return = Dio_ReadPin(2,5, NULL_PTR);
+	UT_ASSERT_EQ("ReadBit_InValidInput", "Dio", Loc_Return, 0,"Correct Return status" ,"Wrong Return status");
+}
```

Comment:
It will be better if you named the test cases to be like the tested function (for example instead of ReadBit, it will be better if it's ReadPin as it will be more clear)